### PR TITLE
Fix plugin control bug from #364

### PR DESCRIPTION
**Pull request by annagrram**
@0xACE This should address the issues you raised.
When you have time, please give it a try.

Thanks!

### CODE REVIEW

**Comment by annagrram** [comment]
Found some problems.
Will write back when they are fixed :)
Sorry about that.

File: src/nnn.c [hunk 1]
```diff
@@ -4935,15 +4935,18 @@ static void browse(char *ipath, const char *session)
...
 
-					mkpath(plugindir, tmp, newpath);
-					if (ndents)
-						spawn(newpath, dents[cur].name, path, path, F_NORMAL);
-					else
-						spawn(newpath, NULL, path, path, F_NORMAL);
+					xstrlcpy(rundir, path, PATH_MAX);
+					xstrlcpy(path, plugindir, PATH_MAX);
+					if (ndents) {
+						if (!run_selected_plugin(&path, tmp, newpath, rundir, dents[cur].name, &lastname, &lastdir))
```
Comment (jarun) [inline src/nnn.c]: (ndents ? dents[cur].name : NULL) will work as a function param too. The less we duplicate code, the better.

Comment (annagrram) [reply src/nnn.c]: True. Thanks!

> The less we duplicate code, the better.

That's the reason I'm still waiting with the fix.
I'm trying to figure out how to do it without code duplication.
I'm a bit busy with something else currently, but a push should be available somewhen today.

Comment (jarun) [reply src/nnn.c]: No hurry. :+1:

Comment (jarun) [reply src/nnn.c]: General question: when using all 4 contexts, how does the `cd` behave? Does it open in the same context from which plugin is run or tries to open a new one?

Comment (jarun) [reply src/nnn.c]: I got it.

**Comment by jarun** [comment]
@annagrram can you please update the PR so we can close it?

**Comment by annagrram** [comment]
Yeah. I haven't forgotten about it.
Will update it this week (hopefully in the next day or two)

**Comment by jarun** [comment]
@annagrram I have incorporated it at commit 53a2e4c56058db01dbe7bcbaeb37a9352fc73959.

**Comment by bimlas** [comment]
Hi!

I'm new here and first of all I like to say thanks for the tool! It really clips into Unix philosophy. :+1:

Thanks for the fix the `fzcd` plugin works, but how do I remove the "Choose context 1-4 (blank for current):" question? Even though I choose the context once, it will ask me again when I use the plugin for the next time. Also, I can only use the plugin in context 1, it does not work in context 2.

**Comment by jarun** [comment]
> it will ask me again when I use the plugin for the next time

That's as per design because you may choose to run the plugin from any context. It doesn't retain the history.

> Also, I can only use the plugin in context 1, it does not work in context 2.

I am assuming you are on master. I can't reproduce this. Please provide the steps.

**Comment by jarun** [comment]
OK, I can reproduce it now.

**Comment by jarun** [comment]
@annagrram please fix this. Invoking a plugin directly with key from any context other than the first one doesn't work.

Also, please check what happens when you do the following:

remove the plugin and use the hotkey. Does `nnn` return to the original directory? Asking this because I see that you modify `path` before calling `run_selected_plugin`. I don't think we should even be doing this because we need to retain these when there's a failure to return to correct path.

**Comment by 0xACE** [comment]
> Thanks for the fix the `fzcd` plugin works, but how do I remove the "Choose context 1-4 (blank for current):" question?

Me personally I have changed `.nnn-plugin-helper` to remove these lines:

```
     if [ -n "$2" ]; then
         context=$2
     else
         echo -n "Choose context 1-4 (blank for current): "
         read context
     fi
```

Which will then default to the current context.

**Comment by jarun** [comment]
Does it work in other contexts still?

**Comment by jarun** [comment]
> Does it work in other contexts still?

I have fixed it now.

I see another issue: the containing directory is not opening correctly.

**Comment by annagrram** [comment]
> @annagrram please fix this. Invoking a plugin directly with key from any context other than the first one doesn't work.
> 
> Also, please check what happens when you do the following:
> 
> remove the plugin and use the hotkey. Does `nnn` return to the original directory? Asking this because I see that you modify `path` before calling `run_selected_plugin`. I don't think we should even be doing this because we need to retain these when there's a failure to return to correct path.

Sorry it took me forever. Difficult month.

I've just pushed a fix, and then I saw you already closed the PR.

**Comment by jarun** [comment]
I've taken care of it. No problem!

Better now?

**Comment by annagrram** [comment]
> I've taken care of it. No problem!
> 
> Better now?

Yeah. Thanks!

Things are starting to improve.

**Comment by jarun** [comment]
Great! :+1:

**Comment by bimlas** [comment]
> I have fixed it now.

I tried, everything works.

> I have changed .nnn-plugin-helper on my local copy to remove these lines ...
> Does it work in other contexts still?

Yes, the modification works fine, I can use the plugin without any problems in any context.

**Comment by jarun** [comment]
Awesome! Thanks for confirming! :+1:

**Pull request closed without merging**